docs/proposals/diamond.md

```diff
@@ -0,0 +1,73 @@
+      1. A node on failure can rerun the same command, as long as the split id is the same
+      the previous filelist contents will be superseded by the ones uploaded later.
+   3. command: ```datamon bundle diamon add --repo {repo} --id {ID} --split-id {client generated ID for each split in a diamond} --source {path}```
+   4. backend model: In the vmetadata bucket the following path will be created or rewritten.
+      ```/diamonds/{repo}/{id}/filelist-{split-id}-{ksuid}.yaml```. KSUID suffix generated at start of the 
+      add command allows for ordering of uploads from the same split.
+   5. This command can also be reused to have incremental uploads before a bundle is created.
+   6. The ordering of ```add``` from the same ```--split-id``` can be strongly ordered based on a design similar to WAL.
+      This is a nice to have addition and can be added later in the implementation.
+1. At the end the file lists are combined into a bundle with the commit message. 
+    1. The eventual bundle has a single namespace and if 2 nodes in the diamond have duplicate files, one will win. 
+    The other files including the winner will be stored in a special path in the bundle. (Proposal ./.{ID}/conflicts/{orig path}/file-suffix of client ID)
+    2. command: ```datamon bundle diamond commit --repo {repo} --id {id} --message {message} --label {label}```
+    3. Optionally a commit can be set to fail on conflict.
+        1. command: ```datamon bundle diamond commit --repo {repo} --id {id} --message {message} --label {label} --no-confict``` 
+    4. A file is written to mark the completion of a diamond and the corresponding bundle ID is written. On failure 2 bundles can be created in the repo.
+    for the same diamond but only one will win closing the bundle. ```/diamonds/repo/{id}/bundle.yaml```
+## Other commands
+ 
+```datamon bundle diamond list --repo {repo}```
+
+```datamon bundle diamond list-files --repo {repo} --id {id}```
+
+## Alternate proposal
+
+### Skip split-id
+The ```split-id``` that is client created can be skipped and replaced by a datamon internally generated KSUID. On retries
+a new id would be used. The final merge at the end will deal with ordering the KSUID and picking a winner if there are 
+conflicts. 
+
+Pros:
+
+1. Client does not need to create or manage ```split-id```
+2. Easy of use (similar to 1)
+
+Cons:
+1. Datamon will loose the mapping of client side logic for splits to actual additions to a diamond. ```split-id``` allows
+the grouping of which files originated from which split. The client can log the output of the command to capture the details.
+
+### User generated diamond id
+
+Diamond ID is used to 
+
+1. Track splits that are grouped together
+2. Separate the namespace of the metadata to allow for concurrent diamonds to the same repo.
+
+Diamond ID can also be client generated which would fail if the same diamond ID is used more than once in the initialize step.
```

Comment:
What is the advantage of having a user-generated diamond id?

Reply:
As an application developer one can choose the scheme you want to use, as an example one can use pipeline name + run id, thus there would not be any need to figure out how to plumb the KSUID through the pipeline. each stage can calculate the same diamond ID. It is not fool proof

Reply:
Got it. For the flood pipeline, we plumb the run id through the pipeline anyway after generating it once so having non-user generated KSUIDs would not be much of a disadvantage

Reply:
@kerneltime @snizamoc once the runID fully becomes part of the datamon specification and metadata (for lineage tracking), we may plan to move further ahead and auto-select this. So adding files to the split could start without any preliminary stage and no explicit client coordination. After discussing that topic with Ritesh, we believe that this could be a nice addition, but we'd like to prove the other concepts before we move into that eventually.

Reply:
Updated, take a look.